packages/node/warpsync/syncing.go

```diff
@@ -0,0 +1,327 @@
+	stateRoot         epoch.MerkleRoot
+	manaRoot          epoch.MerkleRoot
+}
+
+type blockReceived struct {
+	block *tangleold.Block
+	peer  *peer.Peer
+}
+
+func (m *Manager) syncRange(ctx context.Context, start, end epoch.Index, startEC epoch.EC, ecChain map[epoch.Index]epoch.EC, validPeers *set.AdvancedSet[identity.ID]) (err error) {
+	startRange := start + 1
+	endRange := end - 1
+
+	m.startSyncing(startRange, endRange)
+	defer m.endSyncing()
+
+	var wg sync.WaitGroup
+	epochProcessingChan := make(chan epoch.Index)
+	epochProcessingStopChan := make(chan struct{})
+	discardedPeers := set.NewAdvancedSet[identity.ID]()
+
+	// Spawn concurreny amount of goroutines.
+	for worker := 0; worker < m.concurrency; worker++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			for {
+				var targetEpoch epoch.Index
+				select {
+				case <-epochProcessingStopChan:
+					return
+				case targetEpoch = <-epochProcessingChan:
+				}
+
+				success := false
+				for it := validPeers.Iterator(); it.HasNext() && !success; {
+					peerID := it.Next()
+					if discardedPeers.Has(peerID) {
+						m.log.Debugw("skipping discarded peer", "peer", peerID)
+						continue
+					}
+
+					db, _ := database.NewMemDB()
+					tangleTree := smt.NewSparseMerkleTree(db.NewStore(), db.NewStore(), lo.PanicOnErr(blake2b.New256(nil)))
+
+					epochChannels := m.startEpochSyncing(targetEpoch)
+					epochChannels.RLock()
+
+					m.requestEpochBlocks(targetEpoch, ecChain[targetEpoch], peerID)
+
+					dataflow.New(
+						m.epochStartCommand,
+						m.epochBlockCommand,
+						m.epochEndCommand,
+						m.epochVerifyCommand,
+						m.epochProcessBlocksCommand,
+					).WithTerminationCallback(func(params *syncingFlowParams) {
+						epochChannels.RUnlock()
+						m.endEpochSyncing(targetEpoch)
+					}).WithSuccessCallback(func(params *syncingFlowParams) {
+						success = true
+						m.log.Infow("synced epoch", "epoch", params.targetEpoch, "peer", params.peerID)
+					}).WithErrorCallback(func(flowErr error, params *syncingFlowParams) {
+						discardedPeers.Add(params.peerID)
+						m.log.Warnf("error while syncing epoch %d from peer %s: %s", params.targetEpoch, params.peerID, flowErr)
+					}).Run(&syncingFlowParams{
+						ctx:           ctx,
+						targetEpoch:   targetEpoch,
+						targetEC:      ecChain[targetEpoch],
+						targetPrevEC:  ecChain[targetEpoch-1],
+						epochChannels: epochChannels,
+						peerID:        peerID,
+						tangleTree:    tangleTree,
+						epochBlocks:   make(map[tangleold.BlockID]*tangleold.Block),
+					})
+				}
+
+				if !success {
+					err = errors.Errorf("unable to sync epoch %d", targetEpoch)
+				}
+			}
+		}()
+	}
+
+	for ei := startRange; ei <= endRange; ei++ {
+		select {
+		case epochProcessingChan <- ei:
+		case <-ctx.Done():
+		}
+	}
+	close(epochProcessingStopChan)
+
+	wg.Wait()
+	close(epochProcessingChan)
+
+	if err != nil {
+		return errors.Errorf("sync failed for range %d-%d", start, end)
+	}
```

Comment:
Instead of assigning the error in the callback I'd suggest the following: hand over a ctx that you can cancel to the goroutines and make them write into a result channel that contains an error and the completed EI (or maybe we don't even need that and counting is enough). Then you can cancel all goroutines on demand if there's an error and you wait on the result channel  instead of the epochProcessingChan

Reply:
Seems to be addressed now with the resultChan, cancellable errCtx and errorChan.